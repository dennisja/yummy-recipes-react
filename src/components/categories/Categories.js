import React, {Component} from 'react';
import ReactDOM from 'react-dom';
import TimeAgo from 'react-timeago';
import {Link} from 'react-router-dom';
import $ from 'jquery';

import PreLoader, {YummyNotifier} from '../Utilities';
import CategoryRequest from '../../helpers/Categories';

const Category = (props) => {
    const {name, id, owner_details, created, edited} = props.category;

    return (
        <li>
            <div className="collapsible-header">{name}</div>
            <div className="collapsible-body white ">
                <table>
                    <tbody>
                        <tr>
                            <td>
                                Category Name:
                            </td>
                            <td>
                                {name}
                            </td>
                        </tr>
                        <tr>
                            <td>
                                Owner:
                            </td>
                            <td>
                                {`${owner_details.firstname} ${owner_details.lastname}`}
                            </td>
                        </tr>
                        <tr>
                            <td>
                                Created:
                            </td>
                            <td>
                                <TimeAgo date={created+"+3"}/>
                            </td>
                        </tr>
                        <tr>
                            <td>
                                Last Edited:
                            </td>
                            <td>
                                <TimeAgo date={ edited+"+3" }/>
                            </td>
                        </tr>
                    </tbody>
                </table>
                <hr />
                <div className="center-align">
                    <Link to={`edit-category/${id}`} className="btn orange btn-small"><i className="fa fa-edit"/>
                        Edit</Link>
                    <Link to={`delete-category/${id}`} onClick={(e)=>props.deleteCategory(e,id)} className="btn red btn-small"><i className="fa fa-remove"/>
                        Delete</Link>
                </div>
            </div>
        </li>
    )
}

class CategorList extends Component {
    state = {
        categories: [],
    }

    getUserCategories() {
        CategoryRequest
            .fetchUserCategories()
            .then(response => {
                this.setState({categories: response.data.recipe_cats,})
                this.props.onFetchCategories()
            })
            .catch(error => {
                if(error.response){
                    const { status, data } = error.response;
                    //send data and error to a higher order method to signal occurance of an error 
                    //and that no recipe is created yet
                    this.props.handleFetchError(data.errors[0])
                }else if (error.request){
                    console.log(error.request)
                    window.Materialize.toast("Request Can't be made", 5000);
                }
            })
    }

    deleteCategory = (event, id)=>{
        event.preventDefault();
        CategoryRequest.deleteCategory(id)
        .then(response=>{
            //alert user that he has successfully deleted a recipe
            window.Materialize.toast(response.data.message, 5000);
            //update categories list
            this.getUserCategories();
        })
        .catch(error=>{
            if(error.response){
                const { status, data } = error.response;
                window.Materialize.toast(data.errors[0], 5000);
            }else{
                window.Materialize.toast("Request Can't be made", 5000);
            }
        })
    }

    componentWillMount() {
        this.getUserCategories();
    }

    componentDidMount() {
        $('.collapsible').collapsible();
    }

    render() {
        const {categories, fullyLoaded} = this.state;
        // create a list of categories
        const categoryItems = categories.map(category => {
            return (<Category category={category} key={category.id} deleteCategory={this.deleteCategory}/>)
        })


        return (
            <ul className="collapsible popout" data-collapsible="accordion">
                {categoryItems}
            </ul>
        )
    }
}

class Categories extends Component {
    state={
        loadedCategories: false,
        error: false,
        errorMessage: '',
    }

    onFetchCategories = ()=>{
        this.setState({
            loadedCategories: true,
            error: false
        })
    }

    handleFetchError = (errorMessage)=>{
        // alert(errorMessage)
        this.setState({
            errorMessage,
            loadedCategories: false,
            error: true
        })
    }

    render() {
        let loader = null;
        const {loadedCategories, error, errorMessage} = this.state

        if(!loadedCategories && !error){
            loader = <PreLoader message="Fetching Categories........"/>
        }

        if(!loadedCategories && error){
            loader = <YummyNotifier message={errorMessage} />
        }

        return (
            <div className="row">
                <div className="col s4">
                    The side menu will go here
                </div>
                <div className="col s8">
                    <div className="card orange-text center-align">
                        <div className="card-title">Categories</div>
                    </div>
                    {loader}
                    <CategorList onFetchCategories={this.onFetchCategories} handleFetchError={this.handleFetchError}/>
                </div>
            </div>
        )
    }
}

export default Categories;
