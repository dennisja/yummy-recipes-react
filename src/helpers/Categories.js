import Configs from '../configs/Configs';
import Requests from './Requests';

class CategoryRequest extends Requests {
    static addCategory(CategoryRequestData) {
        return CategoryRequest.axiosInstance.post(Configs.api.categories, CategoryRequestData);
    }
    static fetchUserCategories() {
        return CategoryRequest.axiosInstance.get(Configs.api.categories);
    }

    static editCategory(CategoryRequestId, newCategoryRequestData) {
        const { categories } = Configs.api;
        return CategoryRequest.axiosInstance.put(`${categories}${CategoryRequestId}/`, newCategoryRequestData);
    }

    static deleteCategory(CategoryRequestId) {
        const { categories } = Configs.api;
        return CategoryRequest.axiosInstance.delete(`${categories}${CategoryRequestId}/`);
    }

    static getCategory(CategoryRequestId) {
        const { categories } = Configs.api;
        return CategoryRequest.axiosInstance.get(`${categories}${CategoryRequestId}/`);
    }

    static getRecipesInCategory(CategoryRequestId) {
        const { categories, recipesUrl } = Configs.api;
        return CategoryRequest.axiosInstance.get(`${categories}${CategoryRequestId}/${recipesUrl}`);
    }
}

export default CategoryRequest;
