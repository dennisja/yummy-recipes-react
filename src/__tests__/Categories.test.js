import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import Categories, { Category, CategoryList } from '../components/categories/Categories';

const category = {
    name: 'H',
    created: 'Mock Date',
    edited: 'Mock Date',
    id: 1,
    owner_details: {
        firstname: 'H',
        lastname: 'h',
    },
};

describe('Test Categories component', () => {
    const categoriesOutput = shallow(<Categories />);
    it('renders properly', () => {
        expect(shallowToJson(categoriesOutput)).toMatchSnapshot();
    });

    it('renders a sidebar component', () => {
        expect(categoriesOutput.find('SideBar')).toHaveLength(1);
    });

    it('shows preloader when loading categories', () => {
        expect(categoriesOutput.find('PreLoader').length).toBe(1);
        categoriesOutput.setState({ loadedCategories: true });
        expect(categoriesOutput.find('PreLoader').length).toBe(0);
    });
});


describe('Test Category component', () => {
    const categoryWrapper = shallow(<Category category={category} />);
    it('renders properly', () => {
        expect(shallowToJson(categoryWrapper)).toMatchSnapshot();
    });

    it('displays the category name', () => {
        expect(categoryWrapper.find('.collapsible-header').text()).toBe(category.name);
    });

    it('has to Link button elements', () => {
        expect(categoryWrapper.find('Link').length).toBe(2);
    });
});


describe('Test CategoryList component', () => {
    const categoryListWrapper = shallow(<CategoryList />);
    jest.mock('jquery');
    jest.mock('materialize-css');
    it('renders properly', () => {
        expect(shallowToJson(categoryListWrapper)).toMatchSnapshot();
    });

    it('renders categoy elements', () => {
        expect(categoryListWrapper.find('Category')).toHaveLength(0);
    });
});
