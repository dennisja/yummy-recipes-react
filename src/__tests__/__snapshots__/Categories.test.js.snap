// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Categories component renders properly 1`] = `
<div
  className="row"
>
  <div
    className="col s4"
  >
    <SideBar />
  </div>
  <div
    className="col s8"
  >
    <div
      className="card orange-text center-align"
    >
      <div
        className="card-title"
      >
        Categories
      </div>
    </div>
    <PreLoader
      message="Fetching Categories........"
    />
    <CategoryList
      handleFetchError={[Function]}
      onFetchCategories={[Function]}
    />
  </div>
</div>
`;

exports[`Test Category component renders properly 1`] = `
<li>
  <div
    className="collapsible-header"
  >
    H
  </div>
  <div
    className="collapsible-body white "
  >
    <table>
      <tbody>
        <tr>
          <td>
            Category Name:
          </td>
          <td>
            H
          </td>
        </tr>
        <tr>
          <td>
            Owner:
          </td>
          <td>
            H h
          </td>
        </tr>
        <tr>
          <td>
            Created:
          </td>
          <td>
            <TimeAgo
              component="time"
              date="Mock Date+3"
              formatter={[Function]}
              live={true}
              maxPeriod={Infinity}
              minPeriod={0}
              now={[Function]}
            />
          </td>
        </tr>
        <tr>
          <td>
            Last Edited:
          </td>
          <td>
            <TimeAgo
              component="time"
              date="Mock Date+3"
              formatter={[Function]}
              live={true}
              maxPeriod={Infinity}
              minPeriod={0}
              now={[Function]}
            />
          </td>
        </tr>
      </tbody>
    </table>
    <hr />
    <div
      className="center-align"
    >
      <Link
        className="btn orange btn-small"
        replace={false}
        to="edit-category/1"
      >
        <i
          className="fa fa-edit"
        />
        Edit
      </Link>
      <Link
        className="btn red btn-small"
        onClick={[Function]}
        replace={false}
        to="delete-category/1"
      >
        <i
          className="fa fa-remove"
        />
        Delete
      </Link>
    </div>
  </div>
</li>
`;

exports[`Test CategoryList component renders properly 1`] = `
<ul
  className="collapsible popout"
  data-collapsible="accordion"
/>
`;
